### DISCUSSION TITLE
Add Keybinds to change speed modifier of RateAdjust mods

### DISCUSSION THREAD

**Discussion by Fabiano1337** [ppy/osu#28085] [Ideas]
Discussion about PR https://github.com/ppy/osu/pull/28071

Hi,
I would like to add the ability to change your current RateAdjustMod(DoubleTIme for example) speed with 2 keys.

**Example**
(Song Select screen and no mod is selected)

Key1 is pressed
HalfTime mod gets applied with 0.95x speed modifier
Key1 is pressed again
HalfTime mod speed modifier changes to 0.9x
same goes with Key2 and DoubleTime.

If Daycore or Nightcore are currently selected it increases those.
If the minimum rate gets surpassed the opposite mod get applied (HalfTime into DoubleTime and DayCore into NightCore)

**What is the reason for wanting this?**
It is a quality of life feature that makes adjusting the speed of your map faster (useful for people who often push their limits and increase the speed every run of a map), with this implementation you can adjust speed the using only your keyboard.

**Why should DT/HT get preferential hotkey treatment over every other mod in the game?**
DT/HT are some of the most used mods in the game, HR/HD don't have any adjustability. HR and HD can be fully utilized with only a keyboard whereas DT/HT cannot. DC/NC are lesser used than DT/HT which is why i opted for DT/HT, this is up for debate tho.

**I would like your guys feedback**

- Do you want this to be a thing ?
- What keys would be best suited for this ?
- What should happen to WindUp/WindDown

**Comment by NotGumballer91** [comment]
Definitely. But not just limited to RateAdjust mods. Any mods that has any modifier (slider) value with a keypress. Such as difficulty adjust

Since the mod screen is a separate menu from the beatmap list, Arrows keys should suffice? 
Navigating the customization with up and down. Increasing the value with left and right

**Reply by bdach** [reply]
That doesn't work. OP wants to be able to change speed *from song select* using two keys. Not from the mod overlay. 

Suffice to say that every mod is not getting its special key binding. This is not a "keyboard navigation in mod settings customisation menu" proposal.

**Comment by cl8n** [comment]
if all of https://github.com/ppy/osu/pull/28071#issuecomment-2092228261 is addressed then this seems cool & helpful to me

> What keys would be best suited for this ?

ctrl+up ctrl+down as suggested sounds good

> What should happen to WindUp/WindDown

if either of these mods is enabled, make the new keybinds do nothing -- those have more complicated controls and I think it's easy enough to understand that you've opted out of HT/DT stuff when you choose conflicting mods. actually I think that's how it can be implemented too, basically turn off the keybinds if it would create conflicting mods.

**Reply by Fabiano1337** [reply]
I just checked in release and Ctrl+up/Ctrl+down is currently bount to swap between leaderboard types (local,global)

**Reply by longnguyen2004** [reply]
They don't? At least to me they don't do anything (Note: It's Ctrl+Up and Ctrl+Down, not Ctrl+PgUp and Ctrl+PgDn)

**Reply by Fabiano1337** [reply]
It might depend on when you installed osu!lazer and what you chose in the first setup

**Reply by longnguyen2004** [reply]
You're confusing between Ctrl+Up (Up Arrow), which does nothing, with Ctrl+PgUp (Page Up), which switches between the leaderboards like you said.

**Comment by rodrigopina360** [comment]
Honestly I don't think there's much use here. If you want to just play DT/HT it's probably faster to just open the mod select screen and selecting the mods. 
Yeah it could be faster if slightly changing speeds between maps, and I do get that vision there, but an argument can be made for quite a lot of other mods.
Accuracy Challenge, Difficulty Adjust, Wind Down/Wind Up and Adaptive Speed could all be justified, only differing on the amount of people using it (which might or might not change when some of the mods become ranked and/or more people switch to lazer).

I personally wouldn't mind if it was a feature down the line, but unless a more elegant solution is found I don't see a way to make it work that isn't gonna require quite a lot of keybinds (which reminder could grow with new mods down the line).

**Reply by RubiksImplosion** [reply]
I disagree and think this feature would be extremely useful. A convenient button to change a map's rate would save a lot of time and effort. IMO a key for the rate of a map is significantly more useful than adjusting OD/HP or applying more niche mods compared to simple rate changes. The use case I imagine is this: arrow-keying through the carousel until you land on a map you want to play, then pressing the hotkey to adjust the rate to what you want without having to fumble around in the mod select menu, especially now that a rate change involves accessing the rate mod submenu.

**Reply by cl8n** [reply]
> Accuracy Challenge, Difficulty Adjust, Wind Down/Wind Up and Adaptive Speed could all be justified, only differing on the amount of people using it

I mean... yeah. the reason I think a hotkey is more appropriate for rates than any other mod settings is solely due to it being more useful and frequently adjusted. I don't feel as if that will change

but, if you really want to be worried about running out of good default keybinds at some point, there's also the option that they just *don't* have default keybinds, but do allow you to bind them to whatever you want
